### PR TITLE
Merge live ranges for same-bundle aliased pool buffers

### LINKED ISSUE #3980: Compiled decode emits garbage tokens
# torch-spyre#3707: compiled decode emits garbage tokens

❗ ❗ ❗  Reproduced on torch-spyre commit `a3128985`, which is the current pin for spyre-inference

**Symptom.** On the compiled path, greedy decode returns non-tokens instead of
text. `HBM_POOL_PLANNING=0` fixes it. The eager path is unaffected.

**Bisected to** `c9ebe9fd8e7e3f75820cda5d36deb1b46529f3f0` ("Change frontend
codegen to use pool_base_addr to intermediate tensors", #3707) — index 1 in the
range passes, index 2 fails.

This reproducer uses vLLM. We have **not** managed to reproduce the wrong values
with a vLLM-free graph yet- still working on that.

## Reproducer

`repro_garbage_vllm.py`:

```python
"""torch-spyre #3707: compiled greedy decode returns non-tokens.

    HBM_POOL_PLANNING=1 python repro_garbage_vllm.py   # default: garbage
    HBM_POOL_PLANNING=0 python repro_garbage_vllm.py   # expected text

HBM_POOL_PLANNING is read when torch_spyre is imported, so it must come from the
environment -- setting it inside the process after import has no effect.
"""
import os

MODEL = os.environ.get("REPRO_MODEL", "ibm-ai-platform/micro-g3.3-8b-instruct-1b")
PROMPT = "What are IBMs main businesses?"
EXPECTED = "\n\nIBMs main businesses are the companies that provide the services of the"


def main():
    os.environ.setdefault("VLLM_EXECUTE_MODEL_TIMEOUT_SECONDS", "36000")

    from vllm import LLM, SamplingParams
    from torch_spyre._inductor import config

    # enforce_eager=False -> CompilationMode.STOCK_TORCH_COMPILE. The eager path
    # is unaffected by #3707.
    engine = LLM(model=MODEL, enforce_eager=False, max_model_len=128, max_num_seqs=2)
    out = engine.generate(
        PROMPT, SamplingParams(temperature=0.0, max_tokens=16), use_tqdm=False
    )
    text = out[0].outputs[0].text

    print(f"\nhbm_pool_planning={config.hbm_pool_planning}")
    print(f"model={MODEL}")
    print(f"got     ={text!r}")
    print(f"expected={EXPECTED!r}")
    print("VERDICT=" + ("PASS" if text == EXPECTED else "GARBAGE"))


if __name__ == "__main__":
    main()
```

Setup: `spyre-inference` at `a31289852145a59099edccc3e506cf5336e8e2e0`, one
device, `rm -rf /tmp/torchinductor_*` before each run so the inductor FX cache
cannot replay an artifact compiled under the other setting.

## Observed

```
$ HBM_POOL_PLANNING=1 python repro_garbage_vllm.py
hbm_pool_planning=True
model=ibm-ai-platform/micro-g3.3-8b-instruct-1b
got     =',...............'
expected='\n\nIBMs main businesses are the companies that provide the services of the'
VERDICT=GARBAGE

$ HBM_POOL_PLANNING=0 python repro_garbage_vllm.py
hbm_pool_planning=False
model=ibm-ai-platform/micro-g3.3-8b-instruct-1b
got     ='\n\nIBMs main businesses are the companies that provide the services of the'
expected='\n\nIBMs main businesses are the companies that provide the services of the'
VERDICT=PASS
```

The garbage text is **not stable across runs of the same configuration** — the
same setup under pytest produced `'????????????????'`. It degenerates into a
repeated single token either way, i.e. the logits are wrong from the first decode
step, but *which* wrong token wins varies. That points at corruption whose
content depends on what else happens to be resident, rather than a deterministic
miscompile of one kernel.

`google/gemma-3-1b-it` fails the same way (`REPRO_MODEL=google/gemma-3-1b-it`,
expected `"\n\nIBM's main businesses are:\n\n*   **Consulting:** Providing"`).

`SPYRE_COMPILE_GRANULARITY=model` (whole-model graph instead of the default
per-transformer-block graphs) also fails, so this is not specific to the number
of compiled regions.

## Workaround in use

`HBM_POOL_PLANNING=0`, set in the environment before `torch_spyre` is imported
(`torch_spyre/_inductor/config.py:26` reads it at import time).

### PR DESCRIPTION
**Pull request by dgrove-oss**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Fixes a silent buffer-offset clobber in `hbm_pool_planning.py` when two
distinct buffer *names* alias the same underlying storage within a single
SDSC bundle.

`enforce_indirect_access_layout.py`'s `_insert_mutation_relayout_copy`
constructs one `FixedTiledLayout` for a non-compliant scatter's copy-in
buffer, and `propagate_layouts.py`'s `propagate_mutation_layouts` resolves
the retargeted scatter's `MutationLayoutSHOULDREMOVE.real_layout()` to that
*same* layout instance — so two buffer names can share one
`FixedTiledLayout.allocation` dict. `hbm_pool_planning`'s `_is_cross_bundle`
alloc_id guard only excluded aliasing that spans *different* bundles; when
both aliased names were written within the *same* bundle, they were treated
as two independent pool candidates. The allocator then computed two
different offsets and wrote them both into the one shared `.allocation`
dict — the second write silently clobbered the first, corrupting the
buffer's actual device address.

Rather than excluding same-bundle aliased buffers from the pool (simpler,
but overly conservative — mutation buffers of this shape are pervasive in
coarse_tiling, so this would block a lot of otherwise-poolable memory), the
fix extends the live-range analysis so aliased names can still be safely
pool-allocated:

- `_compute_live_ranges` gained a third parameter, `alloc_id_of`, backed by
  the existing `_alloc_id(name) -> id(layout.allocation)` helper. After
  computing each candidate's individual `(start, end)`, a merging pass
  groups names by `alloc_id_of(name)` and overwrites every name's live
  range with the group's union `(min(starts), max(ends))` — aliased names
  are one physical storage location, so their live range must be the union
  of every alias's range, not each in isolation.
- The allocation loop now groups `sorted_bufs` by `alloc_id` and calls
  `allocator.allocate()` exactly once per group (using the merged range),
  writing a single safe offset into the shared `.allocation` dict.
- Per-name codegen bookkeeping (`SpyreEmptyFallback`/`removed_buffers`)
  still runs once for every name in a group, since each name is still
  emitted as its own statement in codegen even though storage is shared.
- Cross-bundle aliasing exclusion (`_is_cross_bundle`) is unchanged — pools
  are bundle-scoped, so aliasing that spans bundles is still excluded
  entirely; only the same-bundle case is now handled via merged live
  ranges instead of being a silent corruption.

#### Which issue(s) this PR is related to:

Related to #3980 (`HBM_POOL_PLANNING=1` causes compiled greedy decode via
vLLM to emit garbage tokens). This same-bundle aliasing bug was found while
investigating #3980 and is a real, independently-reproducible corruption —
but **does not resolve #3980's actual symptom**: re-running the real vLLM
repro with this fix applied still reproduces the garbage-token output.
Ground-truth operand-level tracing across all pool-eligible bundles in a
minimal repro run subsequently ruled out HBM-pool offset reuse entirely as
the cause of #3980 — that investigation is ongoing separately (currently
paused pending a colleague's parallel investigation into the
IndirectAccess/block-table index computation). #3980 should remain open
after this PR merges.

#### Special notes for your reviewer:

- New regression test:
  `test_aliased_buffer_within_same_bundle_shares_one_merged_live_range` in
  `tests/inductor/test_hbm_pool_planning.py`. It proves both aliased names
  become pool-eligible with one shared offset, and that the merge is
  load-bearing: a third, unrelated buffer cannot reuse the aliased pair's
  block while it's still live under the *merged* range, even though it
  would fit under either alias's individual, unmerged range.
- Verification performed:
  - `test_hbm_pool_planning.py`: 20/20 pass.
  - `test_hbm_pool_planning.py` + `test_building_blocks.py` together: 31/31
    pass.
  - Real-hardware `test_coarse_tile_e2e.py` with `HBM_POOL_PLANNING=1`: 188
    passed, 41 skipped, 1 xfailed (pre-existing, unrelated writer-advance
    issue) — no regressions in coarse_tiling's mutation-buffer-heavy
    workloads under pool planning.
  - `pre-commit run --all-files` on the changed files: clean.

#### Does this PR introduce a user-facing change?

No behavior change for callers. `HBM_POOL_PLANNING=1` now pool-allocates a
class of same-bundle aliased mutation buffers that would previously have
silently produced wrong device addresses (a correctness fix, not an API or
default change).

#### Additional note:

This is the first of (at least) two distinct bugs uncovered during the
#3980 investigation. The second — why the real vLLM decode path still
produces garbage tokens under `HBM_POOL_PLANNING=1` even with this fix
applied — remains open and is being tracked separately under #3980 itself.

### CODE REVIEW

**Comment by joerunde** [comment]
@dgrove-oss #3984 causes a very similar output degradation, which I would guess is why this fix cannot be validated against these changes when they're staged on top of main

If we can prove that this change fixes the output when it's staged on top of commit `a31289852145a59099edccc3e506cf5336e8e2e0`, then I think we could close out #3980

**Comment by dgrove-oss** [comment]
@joerunde  -- I was testing against a31289852145a59099edccc3e506cf5336e8e2e0, so although this fixes a real bug in the pool allocator, I'm pretty confident it doesn't fix #3980.  I still get the `VERDICT=GARBAGE` output from the reproducer.

**Review by avery-blanchard — APPROVED** [review]
looks good to me
